File: src/eko-core/prompt/template-engine.ts
/**
 * Enhanced Template Engine for OrbitX
 * Supports nested object access and advanced template features
 */

export interface TemplateContext {
  [key: string]: any
}

export interface TemplateOptions {
  preserveUnresolved?: boolean // 是否保留未解析的占位符
  throwOnMissing?: boolean     // 遇到缺失值时是否抛出错误
}

export class EkoTemplateEngine {
  private static instance: EkoTemplateEngine
  
  /**
   * 获取单例实例
   */
  static getInstance(): EkoTemplateEngine {
    if (!EkoTemplateEngine.instance) {
      EkoTemplateEngine.instance = new EkoTemplateEngine()
    }
    return EkoTemplateEngine.instance
  }

  /**
   * 解析模板占位符，格式为 {PLACEHOLDER} 或 {object.property}
   * @param template 模板字符串
   * @param context 上下文对象
   * @param options 解析选项
   */
  resolve(template: string, context: TemplateContext, options: TemplateOptions = {}): string {
    const { preserveUnresolved = true, throwOnMissing = false } = options

    return template.replace(/\{([^}]+)\}/g, (match, key) => {
      const trimmedKey = key.trim()
      
      try {
        // 支持点号嵌套访问
        const value = this.getNestedValue(context, trimmedKey)
        
        if (value !== undefined && value !== null) {
          return typeof value === 'string' ? value : JSON.stringify(value)
        }
        
        // 处理缺失值
        if (throwOnMissing) {
          throw new Error(`Template placeholder '${trimmedKey}' not found in context`)
        }
        
        // 如果未找到且允许保留，则保留占位符
        return preserveUnresolved ? match : ''
      } catch (error) {
        if (throwOnMissing) {
          throw error
        }
        return preserveUnresolved ? match : ''
      }
    })
  }

  /**
   * 获取嵌套对象的值
   * @param obj 对象
   * @param path 路径，如 'user.profile.name'
   */
  private getNestedValue(obj: any, path: string): any {
    if (!path) return undefined
    
    return path.split('.').reduce((current, key) => {
      if (current === null || current === undefined) {
        return undefined
      }
      
      // 处理数组索引访问，如 items[0]
      const arrayMatch = key.match(/^(\w+)\[(\d+)\]$/)
      if (arrayMatch) {
        const [, arrayKey, index] = arrayMatch
        const array = current[arrayKey]
        return Array.isArray(array) ? array[parseInt(index, 10)] : undefined
      }
      
      return current[key]
    }, obj)
  }

  /**
   * 验证模板中是否包含所有必需的占位符
   * @param template 模板字符串
   * @param requiredPlaceholders 必需的占位符列表
   */
  validate(template: string, requiredPlaceholders: string[]): string[] {
    const missingPlaceholders: string[] = []
    
    for (const placeholder of requiredPlaceholders) {
      const regex = new RegExp(`\\{\\s*${this.escapeRegex(placeholder)}\\s*\\}`, 'g')
      if (!regex.test(template)) {
        missingPlaceholders.push(placeholder)
      }
    }
    
    return missingPlaceholders
  }

  /**
   * 提取模板中的所有占位符
   * @param template 模板字符串
   */
  extractPlaceholders(template: string): string[] {
    const placeholders: string[] = []
    const regex = /\{([^}]+)\}/g
    let match: RegExpExecArray | null
    
    while ((match = regex.exec(template)) !== null) {
      const placeholder = match[1].trim()
      if (!placeholders.includes(placeholder)) {
        placeholders.push(placeholder)
      }
    }
    
    return placeholders
  }

  /**
   * 转义正则表达式特殊字符
   */
  private escapeRegex(str: string): string {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
  }

  /**
   * 批量解析多个模板
   * @param templates 模板对象
   * @param context 上下文对象
   * @param options 解析选项
   */
  resolveMultiple(
    templates: Record<string, string>, 
    context: TemplateContext, 
    options: TemplateOptions = {}
  ): Record<string, string> {
    const result: Record<string, string> = {}
    
    for (const [key, template] of Object.entries(templates)) {
      result[key] = this.resolve(template, context, options)
    }
    
    return result
  }

  /**
   * 检查模板是否有效（语法正确）
   * @param template 模板字符串
   */
  isValidTemplate(template: string): boolean {
    try {
      // 检查括号是否匹配
      let openCount = 0
      for (const char of template) {
        if (char === '{') openCount++
        if (char === '}') openCount--
        if (openCount < 0) return false
      }
      return openCount === 0
    } catch {
      return false
    }
  }
}

/**
 * 便捷函数：快速解析模板
 */
export function resolveTemplate(
  template: string, 
  context: TemplateContext, 
  options?: TemplateOptions
): string {
  return EkoTemplateEngine.getInstance().resolve(template, context, options)
}

/**
 * 便捷函数：提取占位符
 */
export function extractPlaceholders(template: string): string[] {
  return EkoTemplateEngine.getInstance().extractPlaceholders(template)
}

File: src-tauri/src/agent/prompt/template_engine.rs
use regex::Regex;
use serde_json::Value;
use std::collections::HashMap;

/// 模板引擎，支持变量替换和嵌套对象访问
pub struct TemplateEngine {
    // 可以在这里添加缓存等优化
}

impl TemplateEngine {
    pub fn new() -> Self {
        Self {}
    }

    /// 解析模板，支持 {variable} 和 {object.property} 格式
    pub fn resolve(
        &self,
        template: &str,
        context: &HashMap<String, Value>,
    ) -> Result<String, String> {
        let mut result = template.to_string();

        // 使用正则表达式查找所有 {variable} 占位符
        let re = Regex::new(r"\{([^}]+)\}").map_err(|e| e.to_string())?;

        for capture in re.captures_iter(template) {
            let full_match = &capture[0];
            let variable_path = &capture[1].trim();

            // 获取嵌套值
            let value = self.get_nested_value(context, variable_path);

            let replacement = match value {
                Some(Value::String(s)) => s,
                Some(Value::Number(n)) => n.to_string(),
                Some(Value::Bool(b)) => b.to_string(),
                Some(other) => serde_json::to_string(&other).unwrap_or_else(|_| "null".to_string()),
                None => continue, // 保留未找到的占位符
            };

            result = result.replace(full_match, &replacement);
        }

        Ok(result)
    }

    /// 获取嵌套对象的值，支持 "object.property" 格式
    fn get_nested_value(&self, context: &HashMap<String, Value>, path: &str) -> Option<Value> {
        let parts: Vec<&str> = path.split('.').collect();

        if parts.is_empty() {
            return None;
        }

        let mut current = context.get(parts[0])?;

        for part in &parts[1..] {
            current = match current {
                Value::Object(obj) => obj.get(*part)?,
                _ => return None,
            };
        }

        Some(current.clone())
    }

    /// 提取模板中的所有占位符
    pub fn extract_placeholders(&self, template: &str) -> Result<Vec<String>, String> {
        let re = Regex::new(r"\{([^}]+)\}").map_err(|e| e.to_string())?;
        let mut placeholders = Vec::new();

        for capture in re.captures_iter(template) {
            let placeholder = capture[1].trim().to_string();
            if !placeholders.contains(&placeholder) {
                placeholders.push(placeholder);
            }
        }

        Ok(placeholders)
    }

    /// 验证模板语法是否正确
    pub fn validate_template(&self, template: &str) -> bool {
        // 检查括号是否匹配
        let mut open_count = 0;
        for ch in template.chars() {
            match ch {
                '{' => open_count += 1,
                '}' => {
                    open_count -= 1;
                    if open_count < 0 {
                        return false;
                    }
                }
                _ => {}
            }
        }
        open_count == 0
    }
}

impl Default for TemplateEngine {
    fn default() -> Self {
        Self::new()
    }
}
